File: typescript/src/agents/toolCalling/agent.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AgentError, BaseAgent } from '../base.js';
import { AnyTool, DynamicTool, StringToolOutput, ToolError, ToolOutput } from '../../tools/base.js';
import { BaseMemory } from '../../memory/base.js';
import { AgentMeta } from '../types.js';
import { Emitter } from '../../emitter/emitter.js';
import type {
  ToolCallingAgentExecutionConfig,
  ToolCallingAgentTemplates,
  ToolCallingAgentCallbacks,
  ToolCallingAgentRunInput,
  ToolCallingAgentRunOptions,
  ToolCallingAgentRunOutput,
  ToolCallingAgentRunState,
} from './types.js';
import { GetRunContext } from '../../context.js';
import { ChatModel } from '../../backend/chat.js';
import { shallowCopy } from '../../serializer/utils.js';
import { UnconstrainedMemory } from '../../memory/unconstrainedMemory.js';
import { AssistantMessage, SystemMessage, ToolMessage, UserMessage } from '../../backend/message.js';
import { isEmpty, isString } from "remeda";
import { RetryCounter } from '../../internals/helpers/counter.js';
import { mapObj, omitUndefined } from '../../internals/helpers/object.js';
import { Cache } from '../../cache/decoratorCache.js';
import { PromptTemplate } from '../../template.js';
import {
  ToolCallingAgentSystemPrompt,
  ToolCallingAgentTaskPrompt,
} from './prompts.js';
import { z, ZodSchema } from "zod";

export type ToolCallingAgentTemplateFactory<K extends keyof ToolCallingAgentTemplates> = (
  template: ToolCallingAgentTemplates[K],
) => ToolCallingAgentTemplates[K];

export interface ToolCallingAgentInput {
  llm: ChatModel;
  memory: BaseMemory;
  tools: AnyTool[];
  meta?: Omit<AgentMeta, "tools">;
  templates?: Partial<{
    [K in keyof ToolCallingAgentTemplates]:
      | ToolCallingAgentTemplates[K]
      | ToolCallingAgentTemplateFactory<K>;
  }>;
  execution?: ToolCallingAgentExecutionConfig;
  saveIntermediateSteps?: boolean;
}

export class ToolCallingAgent extends BaseAgent<
  ToolCallingAgentRunInput,
  ToolCallingAgentRunOutput,
  ToolCallingAgentRunOptions
> {
  public readonly emitter = Emitter.root.child<ToolCallingAgentCallbacks>({
    namespace: ["agent", "toolCalling"],
    creator: this,
  });

  constructor(public readonly input: ToolCallingAgentInput) {
    super();
    this.input.saveIntermediateSteps = this.input.saveIntermediateSteps ?? true;
  }

  static {
    this.register();
  }

  protected async _run(
    input: ToolCallingAgentRunInput,
    options: ToolCallingAgentRunOptions = {},
    run: GetRunContext<typeof this>,
  ): Promise<ToolCallingAgentRunOutput> {
    const tempMessageKey = "tempMessage" as const;
    const execution = {
      maxRetriesPerStep: 3,
      totalMaxRetries: 20,
      maxIterations: 10,
      ...omitUndefined(this.input.execution ?? {}),
      ...omitUndefined(options.execution ?? {}),
    };

    const state: ToolCallingAgentRunState = {
      memory: new UnconstrainedMemory(),
      result: undefined,
      iteration: 0,
    };
    await state.memory.add(
      new SystemMessage(
        this.templates.system.render({
          role: undefined,
          instructions: undefined,
        }),
      ),
    );
    await state.memory.addMany(this.memory.messages);

    if (input.prompt) {
      const userMessage = new UserMessage(
        this.templates.task.render({
          prompt: input.prompt,
          context: input.context,
          expectedOutput: isString(input.expectedOutput) ? input.expectedOutput : undefined,
        }),
      );
      await state.memory.add(userMessage);
    }

    const globalRetriesCounter = new RetryCounter(execution.totalMaxRetries || 1, AgentError);

    const usePlainResponse = !input.expectedOutput || !(input.expectedOutput instanceof ZodSchema);
    const finalAnswerToolSchema = usePlainResponse
      ? z.object({
          response: z.string().describe(String(input.expectedOutput ?? "")),
        })
      : (input.expectedOutput as ZodSchema);

    const finalAnswerTool = new DynamicTool({
      name: "final_answer",
      description: "Sends the final answer to the user",
      inputSchema: finalAnswerToolSchema,
      handler: async (input) => {
        const result = usePlainResponse ? input.response : JSON.stringify(input.response);
        state.result = new AssistantMessage(result);
        return new StringToolOutput("Message has been sent");
      },
    });

    const tools = [...this.input.tools, finalAnswerTool];
    let forceFinalAnswer = false;

    while (!state.result) {
      state.iteration++;
      if (state.iteration > (execution.totalMaxRetries ?? Infinity)) {
        throw new AgentError(
          `Agent was not able to resolve the task in ${state.iteration} iterations.`,
        );
      }

      await run.emitter.emit("start", { state });
      const response = await this.input.llm.create({
        messages: state.memory.messages.slice(),
        tools,
        toolChoice: forceFinalAnswer ? finalAnswerTool : tools.length > 1 ? "required" : tools[0],
        stream: false,
      });
      await state.memory.addMany(response.messages);

      const toolCallMessages = response.getToolCalls();
      for (const toolCall of toolCallMessages) {
        try {
          const tool = tools.find((tool) => tool.name === toolCall.toolName);
          if (!tool) {
            throw new AgentError(`Tool ${toolCall.toolName} does not exist!`);
          }

          const toolInput: any = toolCall.args;
          const toolResponse: ToolOutput = await tool.run(toolInput).context({
            state,
            toolCallMsg: toolCall,
          });
          await state.memory.add(
            new ToolMessage({
              type: "tool-result",
              toolCallId: toolCall.toolCallId,
              toolName: toolCall.toolName,
              result: toolResponse.getTextContent(),
              isError: false,
            }),
          );
        } catch (e) {
          if (e instanceof ToolError) {
            globalRetriesCounter.use(e);
            await state.memory.add(
              new ToolMessage({
                type: "tool-result",
                toolCallId: toolCall.toolCallId,
                toolName: toolCall.toolName,
                result: e.explain(),
                isError: true,
              }),
            );
          } else {
            throw e;
          }
        }
      }

      // handle empty messages for some models
      const textMessages = response.getTextMessages();
      if (isEmpty(toolCallMessages) && isEmpty(textMessages)) {
        await state.memory.add(new AssistantMessage("\n", { [tempMessageKey]: true }));
      } else {
        await state.memory.deleteMany(
          state.memory.messages.filter((msg) => msg.meta[tempMessageKey]),
        );
      }

      // Fallback for providers that do not support structured outputs
      if (!isEmpty(textMessages) && isEmpty(toolCallMessages)) {
        forceFinalAnswer = true;
        tools.length = 0;
        tools.push(finalAnswerTool);
      }

      await run.emitter.emit("success", { state });
    }

    if (this.input.saveIntermediateSteps) {
      this.memory.reset();
      await this.memory.addMany(state.memory.messages.slice(1));
    } else {
      await this.memory.addMany(state.memory.messages.slice(-2));
    }
    return { memory: state.memory, result: state.result };
  }

  get meta(): AgentMeta {
    const tools = this.input.tools.slice();

    if (this.input.meta) {
      return { ...this.input.meta, tools };
    }

    return {
      name: "ToolCalling",
      tools,
      description: "ToolCallingAgent that uses tools to accomplish the task.",
      ...(tools.length > 0 && {
        extraDescription: [
          `Tools that I can use to accomplish given task.`,
          ...tools.map((tool) => `Tool '${tool.name}': ${tool.description}.`),
        ].join("\n"),
      }),
    };
  }

  @Cache({ enumerable: false })
  protected get templates(): ToolCallingAgentTemplates {
    const overrides = this.input.templates ?? {};
    const defaultTemplates: ToolCallingAgentTemplates = {
      system: ToolCallingAgentSystemPrompt,
      task: ToolCallingAgentTaskPrompt,
    } as const;

    return mapObj(defaultTemplates)(
      (key, defaultTemplate: ToolCallingAgentTemplates[typeof key]) => {
        const override = overrides[key] ?? defaultTemplate;
        if (override instanceof PromptTemplate) {
          return override;
        }
        return override(defaultTemplate) ?? defaultTemplate;
      },
    );
  }

  createSnapshot() {
    return {
      ...super.createSnapshot(),
      input: shallowCopy(this.input),
      emitter: this.emitter,
    };
  }

  set memory(memory: BaseMemory) {
    this.input.memory = memory;
  }

  get memory() {
    return this.input.memory;
  }
}

File: python/beeai_framework/agents/tool_calling/agent.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from collections.abc import Sequence
from typing import Any

from pydantic import BaseModel

from beeai_framework.agents import AgentError, AgentExecutionConfig
from beeai_framework.agents.base import BaseAgent
from beeai_framework.agents.tool_calling.events import (
    ToolCallingAgentStartEvent,
    ToolCallingAgentSuccessEvent,
    tool_calling_agent_event_types,
)
from beeai_framework.agents.tool_calling.prompts import ToolCallingAgentTaskPromptInput
from beeai_framework.agents.tool_calling.types import (
    ToolCallingAgentRunOutput,
    ToolCallingAgentRunState,
    ToolCallingAgentTemplateFactory,
    ToolCallingAgentTemplates,
    ToolCallingAgentTemplatesKeys,
)
from beeai_framework.backend.chat import ChatModel
from beeai_framework.backend.message import (
    AssistantMessage,
    MessageToolResultContent,
    SystemMessage,
    ToolMessage,
    UserMessage,
)
from beeai_framework.context import Run, RunContext
from beeai_framework.emitter import Emitter
from beeai_framework.memory.base_memory import BaseMemory
from beeai_framework.memory.unconstrained_memory import UnconstrainedMemory
from beeai_framework.template import PromptTemplate
from beeai_framework.tools import ToolError
from beeai_framework.tools.tool import AnyTool
from beeai_framework.utils.counter import RetryCounter
from beeai_framework.utils.strings import to_json

__all__ = ["ToolCallingAgent"]


class ToolCallingAgent(BaseAgent[ToolCallingAgentRunOutput]):
    def __init__(
        self,
        *,
        llm: ChatModel,
        memory: BaseMemory | None = None,
        tools: Sequence[AnyTool] | None = None,
        templates: dict[ToolCallingAgentTemplatesKeys, PromptTemplate[Any] | ToolCallingAgentTemplateFactory]
        | None = None,
    ) -> None:
        super().__init__()
        self._llm = llm
        self._memory = memory or UnconstrainedMemory()
        self._tools = tools or []
        self._templates = self._generate_templates(templates)

    def run(
        self,
        prompt: str | None = None,
        *,
        context: str | None = None,
        expected_output: str | type[BaseModel] | None = None,
        execution: AgentExecutionConfig | None = None,
    ) -> Run[ToolCallingAgentRunOutput]:
        execution_config = execution or AgentExecutionConfig(
            max_retries_per_step=3,
            total_max_retries=20,
            max_iterations=10,
        )

        async def handler(run_context: RunContext) -> ToolCallingAgentRunOutput:
            state = ToolCallingAgentRunState(memory=UnconstrainedMemory(), result=None, iteration=0)
            await state.memory.add(SystemMessage(self._templates.system.render()))
            await state.memory.add_many(self.memory.messages)

            if prompt:
                task_input = ToolCallingAgentTaskPromptInput(
                    prompt=prompt,
                    context=context,
                    expected_output=expected_output if isinstance(expected_output, str) else None,
                )
                user_message = UserMessage(self._templates.task.render(task_input))
                await state.memory.add(user_message)

            global_retries_counter = RetryCounter(
                error_type=AgentError, max_retries=execution_config.total_max_retries or 1
            )

            while state.result is None:
                state.iteration += 1

                if execution_config.max_iterations and state.iteration > execution_config.max_iterations:
                    raise AgentError(f"Agent was not able to resolve the task in {state.iteration} iterations.")

                await run_context.emitter.emit(
                    "start",
                    ToolCallingAgentStartEvent(state=state),
                )
                response = await self._llm.create(messages=state.memory.messages, tools=list(self._tools), stream=False)
                await state.memory.add_many(response.messages)

                tool_call_messages = response.get_tool_calls()
                for tool_call in tool_call_messages:
                    try:
                        tool = next((tool for tool in self._tools if tool.name == tool_call.tool_name), None)
                        if not tool:
                            raise ToolError(f"Tool '{tool_call.tool_name}' does not exist!")

                        tool_input = json.loads(tool_call.args)
                        tool_response = await tool.run(tool_input).context(
                            {"state": state.model_dump(), "tool_call_msg": tool_call}
                        )
                        await state.memory.add(
                            ToolMessage(
                                MessageToolResultContent(
                                    result=tool_response.get_text_content(),
                                    tool_name=tool_call.tool_name,
                                    tool_call_id=tool_call.id,
                                )
                            )
                        )
                    except ToolError as e:
                        global_retries_counter.use(e)
                        await state.memory.add(
                            ToolMessage(
                                MessageToolResultContent(
                                    result=self._templates.tool_error.render({"reason": e.explain()}),
                                    tool_name=tool_call.tool_name,
                                    tool_call_id=tool_call.id,
                                )
                            )
                        )

                # handle empty messages for some models
                text_messages = response.get_text_messages()
                if not tool_call_messages and not text_messages:
                    await state.memory.add(AssistantMessage("\n", {"tempMessage": True}))
                else:
                    await state.memory.delete_many(
                        [msg for msg in state.memory.messages if msg.meta.get("tempMessage", False)]
                    )

                if text_messages and not tool_call_messages:
                    if (
                        expected_output is not None
                        and isinstance(expected_output, type)
                        and issubclass(expected_output, BaseModel)
                    ):
                        structured = await self._llm.create_structure(
                            schema=expected_output, messages=state.memory.messages
                        )
                        state.result = AssistantMessage(
                            content=to_json(
                                expected_output.model_validate(structured.object),
                                indent=4,
                            )
                        )
                    else:
                        state.result = AssistantMessage.from_chunks(text_messages)

                await run_context.emitter.emit(
                    "success",
                    ToolCallingAgentSuccessEvent(state=state),
                )

            await self.memory.add_many(state.memory.messages[1:])
            return ToolCallingAgentRunOutput(result=state.result, memory=state.memory)

        return self._to_run(handler, signal=None, run_params={"prompt": prompt, "execution": execution})

    def _create_emitter(self) -> Emitter:
        return Emitter.root().child(
            namespace=["agent", "tool_calling"], creator=self, events=tool_calling_agent_event_types
        )

    @property
    def memory(self) -> BaseMemory:
        return self._memory

    @memory.setter
    def memory(self, memory: BaseMemory) -> None:
        self._memory = memory

    @staticmethod
    def _generate_templates(
        overrides: dict[ToolCallingAgentTemplatesKeys, PromptTemplate[Any] | ToolCallingAgentTemplateFactory]
        | None = None,
    ) -> ToolCallingAgentTemplates:
        templates = ToolCallingAgentTemplates()
        if overrides is None:
            return templates

        for name, _info in ToolCallingAgentTemplates.model_fields.items():
            override: PromptTemplate[Any] | ToolCallingAgentTemplateFactory | None = overrides.get(name)
            if override is None:
                continue
            elif isinstance(override, PromptTemplate):
                setattr(templates, name, override)
            else:
                setattr(templates, name, override(getattr(templates, name)))
        return templates
